### PR TITLE
Fix URL regexp

### LINKED ISSUE #4143: Bad URL detection in Description field
- Gitea version (or commit ref): fa3ceb6 (try.gitea.io)
- Git version: (not relevant)
- Operating system: (unknown)
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes - https://try.gitea.io/netvor/slop-copr-build
  - [ ] No
  - [ ] Not relevant
- Log gist:

## Description

If I include URL inside parentheses in Description of a git repo, gitea web UI will linkify it, but the closing bracket will be detected as  art of the URL, resulting in HTTP-404 errors.

## Screenshots

http://file.vornet.cz/tmp/rbrack.png

### PR DESCRIPTION
**Pull request by axifive**
fix #4143

### CODE REVIEW

**Comment by lunny** [comment]
@axifive CI failed.

**Comment by axifive** [comment]
@lunny I fixed error in the test. Can you run drone again?

**Comment by daviian** [comment]
@axifive Another problem could be that your are expecting a string instead of the `template.HTML`

**Comment by axifive** [comment]
@daviian I compare by value, `template.HTML` has `string` base type

**Comment by appleboy** [comment]
@axifive Could you also add more testing? like `http://user@xxx.com/`

File: models/repo.go [hunk 1]
```diff
@@ -783,7 +783,7 @@ func (repo *Repository) NextIssueIndex() int64 {
 }
 
 var (
-	descPattern = regexp.MustCompile(`https?://\S+`)
+	descPattern = regexp.MustCompile(`https?://[a-zA-Z0-9-_@:;%.,+~#?&/=|!$]+`)
```
Comment (HaasJona) [inline models/repo.go:786]: I don't think, that that is ideal. For example https://www.rügen.de is a valid domain. Also dots, exclamation marks and so on at the end of the url should probably not be included. Just like github doesn't link the punctuation here: http://example.org... or here: http://example.org?

**Comment by axifive** [comment]
@appleboy Ok, I'll add tests after solving the detected problems with the expression

File: models/repo.go (hunk 1 above)
Comment (axifive) [reply models/repo.go:786]: Interesting, GitHub doesn't have an URL detection in the description: https://github.com/axifive/test

**Comment by tonivj5** [comment]
Is this PR useful yet?

**Pull request closed without merging**